FEATURE: validate app manifest

rpl opened the pull request:
This change introduce a validation step in the install/update pipeline
and show validation errors descriptions into the dashboard,
a new remote SimulatorActor command (validateManifest) use
b2g PermissionsTable and AppsUtils.checkManifest to detect
and prevents known error conditions.
- update cached manifest
- validate manifest on install/update

blocking errors:
- missing manifest
- manifest json parsing error (using jsonlint)

non-blocking errors: 
- missing name
- missing icons (fix #311)
- missing 128 icon (needed for marketplace submission)
- wrong type (e.g. hosted could not be type privileged/certified)
- unkown permission
- denied permission
- invalid access

simulator warnings:
- certified apps are not officially supported on the simulator
- WebAPI 'WebSMS' is not currently supported on the simulator
- WebAPI 'WebTelephony' is not currently supported on the simulator

DEMO: http://youtu.be/Py8FCadnOH8

rpl committed:
prototyping app validation helper

rpl committed:
add update cached manifest helper and collect validation errors during validation

rpl committed:
automatically validate manifest on install/update app and visualize validation errors on dashboard

rpl committed:
remove trailing spaces

rpl committed:
FIX: tollerate missing name in manifest, removed dead code and hide validation errors on removed app

rpl committed:
prevent install on manifest missing or parsing error + jsonlint readable error messages on manifest parsing error

rpl committed:
fixed typos and added comments

rpl committed:
removing trailing spaces

rpl committed:
tollerate missing icons and add non-blocking validation error

mykmelez commented on addon/lib/jsonlint.js:
```diff
@@ -0,0 +1,419 @@
+/* jsonlint.js from https://github.com/zaach/jsonlint */
```
Instead of adding this line, which makes the source file different from upstream, add zaach to the list of contributors (alphabetized by first name) in addon/package.json.

Then add a LICENSE file to the top-level directory of the repository, and put the following into it:

All of the source code to this product is available under licenses that are both
free and open source.  Most of it is available under the Mozilla Public License
2.0 (MPL), as specified by the license headers in each file.

The file addon/lib/jsonlint.js is available under the following license:

MIT License

Copyright (C) 2012 Zachary Carter

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

mykmelez replied:
Note: I added a LICENSE file as part of the "push to device" changes, so this would just need to be added to that file.

mykmelez commented on addon/lib/jsonlint.js:
```diff
@@ -0,0 +1,419 @@
+exports.parser = jsonlint;
+exports.parse = function () { return jsonlint.parse.apply(jsonlint, arguments); }
+}
\ No newline at end of file
```
Nit: add newline at end of file.

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -590,6 +599,122 @@ let simulator = module.exports = {
+      Request({
+        url: id,
+        onComplete: function (response) {
+          let error = null
```
Nit: append semicolon.

Also, since error is always set before being used, you don't need to initialize it, so it's preferable to simply declare it, since that is simpler, and it also means that you never change its type (from object to string).

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -590,6 +599,122 @@ let simulator = module.exports = {
+        onComplete: function (response) {
+          let error = null
+          if (response.status != 200) {
+            error = "Unexpected status code " + response.status;
```
Nit: for consistency, format this the same as the "unexpected content-type" error wrt. the colon and trailing period.

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -590,6 +599,122 @@ let simulator = module.exports = {
+
+    this._updateCachedManifest(id, function(error, manifest) {
+      if (error) {
+        app.validation.errors.push("error updating cached manifest: "+error);
```
Nit: add space around plus sign.

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -590,6 +599,122 @@ let simulator = module.exports = {
+            app.manifest = response.json;
+            let contentType = response.headers["Content-Type"];
+            if (contentType !== "application/x-web-app-manifest+json") {
+              error = "Unexpected Content-Type: " + contentType + ".";
```
Nit: all the errors in validateApp start with a lower-case letter, but the ones here in _updateCachedManifest start with an upper-case one. Seems like they should be consistent.

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -590,6 +599,122 @@ let simulator = module.exports = {
+
+      if (["generated", "hosted"].indexOf(app.type) !== -1 &&
+          ["certified", "privileged"].indexOf(app.manifest.type) !== -1) {
+        app.validation.errors.push("hosted app could not be type '"+app.manifest.type+"'");
```
Nit: could not -> can't

mykmelez replied:
Nit: add space around plus signs.

mykmelez commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -186,6 +186,76 @@ SimulatorActor.prototype = {
+        validation: { errors: ["AppsUtils.checkManifest return false"] },
+        error: "apputils-checkmanifest-invalid"
+      }
+    }
```
It seems like app type, checkManifest(), and permissions errors can all coexist, so it'd be better to collect the first two kinds of errors with the permissions errors and return all of them at once instead of exiting early when we detect one of the first two kinds.

mykmelez commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -186,6 +186,76 @@ SimulatorActor.prototype = {
+        message: "Invalid Manifest",
+        validation: { errors: ["Unknown app type: " + appType] },
+        error: "unknown-app-type"
+      }
```
Nit: append semi-colon.

mykmelez commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -186,6 +186,76 @@ SimulatorActor.prototype = {
+      }
+    }
+
+    let valid = AppsUtils.checkManifest(manifest,{});
```
Nit: put space between comma and object literal:

```
let valid = AppsUtils.checkManifest(manifest, {});
```

mykmelez commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -186,6 +186,76 @@ SimulatorActor.prototype = {
+    Cu.import("resource://gre/modules/PermissionsTable.jsm", utils);
+    Cu.import("resource://gre/modules/AppsUtils.jsm", utils);
+
+    let AppsUtils = utils.AppsUtils;
```
Nit: since you only use _AppsUtils_ once, it would be simpler and hence more readable to dereference it at the call point rather than assigning it to an intermediary variable, just like you do with _PermissionsTable_, i.e.:

```
let valid = utils.AppsUtils.checkManifest(manifest, {});
```

mykmelez commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -186,6 +186,76 @@ SimulatorActor.prototype = {
+      let permission = utils.PermissionsTable[name];
+
+      if (permission) {
+        let permission_action = permission[appType];
```
Nit: permission_action = permissionAction for consistency with naming convention.

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -590,6 +599,122 @@ let simulator = module.exports = {
+                  concat(reply.validation.errors);
+              }
+              if (typeof next === "function") {
+                // NOTE: non blocking errors
```
A bit more context would be useful here: why are they non-blocking errors?

mykmelez commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -186,6 +186,76 @@ SimulatorActor.prototype = {
+
+    return {
+      success: true,
+      message: "No errors detected in your manifest"
```
This message is never used, so it seems like we should leave it out.

mykmelez replied:
To elaborate a bit: the _message_ property is never used, whatever the outcome of validation; and it repeats the information available in the _success_ property, since it's only ever "Invalid Manifest" or ""No errors detected in your manifest"; so we should leave it out and use the value of _success_ on the other side to determine whether or not validation was successful (and display a message accordingly).

mykmelez commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -186,6 +186,76 @@ SimulatorActor.prototype = {
+        success: false,
+        message: "Invalid Manifest",
+        validation: { errors: ["Unknown app type: " + appType] },
+        error: "unknown-app-type"
```
This and the other _error_ codes are never used, unlike the validation errors, so it seems like we should leave them out.

mykmelez commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -186,6 +186,76 @@ SimulatorActor.prototype = {
+      return {
+        success: false,
+        message: "Invalid Manifest",
+        validation: { errors: ["Unknown app type: " + appType] },
```
This and the other responses in onValidateManifest are already explicitly identified validation responses, so it seems unnecessary to put the list of errors inside a _validation_ property. I would simply have a top-level _errors_ property.

mykmelez commented:
To wrap up all the various comments I made… it seems like the response to a validation request can be as simple as:

```
{
  success: (Boolean),
  errors: [ … ]
}
```

mykmelez commented:
Or you could even leave out _success_, since validation succeeds if there aren't any errors and fails if there are.

rpl committed:
move jsonlint into its own dir (addon/packages/jsonlint)

- added LICENSE file
- tweaks on package.json to load correctly into cfx
- tweaks on lib/jsonlint.js to load correctly into cfx

rpl committed:
fixed onValidateManifest and validation error reporting messages

rpl committed:
put jsonlint error messages inside a <pre>

rpl committed:
fix error messages, reduce redundant error reporing and detect b2g executable not found

- remove comments about removed app.xid attribute

NOTE: report error on b2g executable not found
(e.g. corrupted installation, running firefox 32bit on a 64bit system, ...)

rpl committed:
remove trailing spaces

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -599,6 +615,139 @@ let simulator = module.exports = {
+                  concat(reply.errors);
+              }
+              if (typeof next === "function") {
+                // NOTE: non blocking errors (does not corrupt the b2g profile)
```
Hmm, it sounds like we define "blocking error" as "corrupts the B2G profile". But that isn't what actually matters to app developers, whose primary concern is whether or not they can test their apps. And in any case profile corruption is the responsibility of the Simulator, not the app developer. It represents a bug in the Simulator rather than a problem with the app developer's manifest (whether or not the manifest is valid).

So I think we should focus the "blocking error" category on issues that prevent an app from being tested in the Simulator, whether or not the issues cause profile corruption. The "non-blocking error" category (which seems analogous to a "warning" in many error-reporting systems) can then represent other kinds of problems, like those that reduce the effectiveness of testing (no icon) or might prevent the developer from uploading their app to the Marketplace (which requires at least one icon that is 128x128, according to the [app manifest docs](https://developer.mozilla.org/en-US/docs/Apps/Manifest#icons)). Although I'm cautious about adding warnings, which often are more trouble than they're worth and cause developers to tune out all messages.

rpl committed:
Move jsonlint into git submodule (custom addon-sdk branch)

rpl committed:
fix bug on validate manifest simualtorActor command

+ refactoring out permission table validation into a separate helper method

rpl committed:
tweaks on validateApp helper + added simulator supports warnings

- describe validation policy in the validateApp commend
- added non-blocking error missing 128 icon (needed for marketplace submission)
- added warnings (warning users when manifest contains something not fully supported
  by the simulator)

rpl commented:
In e5baeab I've added "simulator warning":
when a manifest is formally correct but app permissions or type are not currently supported on the simulator. 

App with warnings:

![app_warnings](https://f.cloud.github.com/assets/11484/211182/08177dd2-82ac-11e2-8550-e867cf07d821.png)

App with warning and errors:

![app_errors_and_warnings](https://f.cloud.github.com/assets/11484/211185/19ee2c5e-82ac-11e2-8bbd-7fa246efc17e.png)

rpl commented:
jsonlint moved in its own submodule: 
- https://github.com/rpl/jsonlint/branches/addon-sdk

rpl committed:
tweaks on validateApp

- non-blocking errors are now warnings
- missing name is now a blocking error
- renamed generateNotSupportedWebAPIWarnings to _validateWebAPIs

rpl committed:
fixed typos

rpl committed:
tweaks on generated validations html fragment

rpl committed:
removed trailing spaces

rpl committed:
removed unused return value from _validateWebAPIs private helper

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -599,6 +611,194 @@ let simulator = module.exports = {
+    });
+  },
+
+  _validateWebAPIs: function(manifest, errors, warnings) {
```
Erm, _errors_ and _warnings_ should be the first two parameters to validation helpers like this one, since the rest of the parameters will vary by helper, and that way the positions of _errors_ and _warnings_ will stay the same across all such functions.

rpl committed:
tweaks on _validateWebAPIs (manifest as last parameter)

rpl commented:
@mykmelez what about if I extract a _validateNameIcons helper from validateApp?

mykmelez commented:
> @mykmelez what about if I extract a _validateNameIcons helper from validateApp?

Great idea!

rpl committed:
extract _validateNameIcons helper from validateApp

rpl committed:
change AppUtils.checkManifest validation error message to be more meaningful

mykmelez commented:
Looks great! I may want to massage the messages a little, but that shouldn't prevent this from landing. Better to land it and then iterate!
